pydantic/_internal/_utils.py

```diff
@@ -387,3 +390,35 @@ def __getitem__(self, key: str, /) -> Any:
 
         def __contains__(self, key: str, /) -> bool:
             return self.wrapped.__contains__(key)
+
+
+_ModelT = TypeVar('_ModelT', bound='BaseModel')
+_R = TypeVar('_R')
+
+
+class deprecatedinstanceproperty(Generic[_ModelT, _R]):
+    """A decorator exposing the decorated class method as a property, with a warning on instance access.
+
+    This decorator takes a class method defined on the `BaseModel` class and transforms it into
+    an attribute. The attribute can be accessed on both the class and instances of the class. If accessed
+    via an instance, a deprecation warning is emitted stating that instance access will be removed in V3.
+    """
+
+    def __init__(self, fget: Callable[[type[_ModelT]], _R], /) -> None:
+        self.fget = fget
+
+    @overload
+    def __get__(self, instance: None, objtype: type[_ModelT]) -> _R: ...
+    @overload
+    @deprecated(
+        'Accessing this attribute on the instance is deprecated, and will be removed in Pydantic V3.',
```

Comment:
Maybe add a note about proper access on the class to guide / assist users?

Reply:
Like you've done below:

Instead, you should access this attribute from the model class.